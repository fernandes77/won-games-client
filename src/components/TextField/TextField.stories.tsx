import type { Meta, StoryObj } from '@storybook/react'
import { IconMail } from '@tabler/icons-react'
import { TextField } from './TextField'

const meta: Meta<typeof TextField> = {
  title: 'Form/TextField',
  component: TextField
}

export default meta
type Story = StoryObj<typeof TextField>

export const Primary: Story = {
  decorators: [
    (Story) => (
      <div style={{ maxWidth: 300, padding: 15 }}>
        <Story />
      </div>
    )
  ],
  args: {
    icon: <IconMail />,
    label: 'E-mail',
    name: 'email',
    id: 'Email',
    initialValue: '',
    placeholder: 'john.cage@gmail.com',
    disabled: false
  }
}
