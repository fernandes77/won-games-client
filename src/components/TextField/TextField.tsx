import { useState } from 'react'

export type TextFieldProps = {
  onInput?: (value: string) => void
  label?: string
  labelFor?: string
  initialValue?: string
  icon?: React.ReactNode
  iconPosition?: 'left' | 'right'
  disabled?: boolean
  error?: string
} & React.InputHTMLAttributes<HTMLInputElement>

export const TextField = ({
  icon,
  iconPosition = 'left',
  label,
  labelFor = '',
  initialValue = '',
  error,
  disabled = false,
  onInput,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInput && onInput(newValue)
  }

  const cssIconPosition = { left: 'order-none', right: 'order-1' }
  const cssIconPositionInput = { left: 'pl-xs', right: 'pr-xs' }

  const cssDisabled = 'cursor-not-allowed text-gray placeholder:text-current'

  return (
    <div>
      {!!label && (
        <label
          htmlFor={labelFor}
          className={`${disabled ? cssDisabled : ''}
          ${error ? 'text-red' : ''}
          text-sm text-black cursor-pointer
          `}
        >
          {label}
        </label>
      )}
      <div
        className={`${error ? 'border-red' : ''}
        flex bg-gray-light rounded-sm px-xs border-2 border-solid border-gray-light
        focus-within:shadow-[0_0_0.5rem] focus-within:shadow-primary`}
      >
        {!!icon && (
          <div
            className={`${cssIconPosition[iconPosition]}
            ${disabled ? cssDisabled : ''}
            ${error ? 'text-red' : ''}
            flex items-center w-[1.375rem] text-gray [&>svg]:w-full`}
          >
            {icon}
          </div>
        )}
        <input
          className={`${cssIconPositionInput[iconPosition]}
          ${disabled ? cssDisabled : ''}
           text-black font-poppins text-base py-xxs bg-transparent border-0 outline-none w-full`}
          type="text"
          onChange={onChange}
          value={value}
          disabled={disabled}
          {...props}
        />
      </div>
      {!!error && <p className="text-red text-xs">{error}</p>}
    </div>
  )
}
