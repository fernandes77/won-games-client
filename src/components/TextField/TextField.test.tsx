import { render, screen, user, waitFor } from '@/utils/tests/helpers'
import { IconMail } from '@tabler/icons-react'
import { TextField } from './TextField'

describe('<TextField />', () => {
  it('Renders with Label', () => {
    render(<TextField label="Label" name="Label" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    render(<TextField />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    render(<TextField placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Renders with Icon', () => {
    render(<TextField icon={<IconMail data-testid="icon" />} />)

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('Renders with Icon on the right side', () => {
    render(
      <TextField icon={<IconMail data-testid="icon" />} iconPosition="right" />
    )

    expect(screen.getByTestId('icon').parentElement).toHaveClass('order-1')
  })

  it('Changes its value when typing', async () => {
    const onInput = jest.fn()
    render(<TextField onInput={onInput} label="TextField" name="TextField" />)

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    await user.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('Does not changes its value when disabled', async () => {
    const onInput = jest.fn()
    render(
      <TextField
        onInput={onInput}
        label="TextField"
        name="TextField"
        disabled
      />
    )

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my new text'
    user.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onInput).not.toHaveBeenCalled()
  })

  it('Renders with error', () => {
    const { container } = render(
      <TextField
        icon={<IconMail data-testid="icon" />}
        label="TextField"
        error="Error message"
      />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', async () => {
    render(<TextField label="TextField" name="TextField" />)

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    await user.tab()
    expect(input).toHaveFocus()
  })

  it('Is not accessible by tab when disabled', () => {
    render(<TextField label="TextField" name="TextField" disabled />)

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    user.tab()
    expect(input).not.toHaveFocus()
  })
})
