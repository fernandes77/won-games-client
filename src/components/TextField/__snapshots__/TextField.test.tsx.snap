// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Renders with error 1`] = `
<div>
  <label
    class="
          text-red
          text-sm text-black cursor-pointer
          "
  >
    TextField
  </label>
  <div
    class="border-red
        flex bg-gray-light rounded-sm px-xs border-2 border-solid border-gray-light
        focus-within:shadow-[0_0_0.5rem] focus-within:shadow-primary"
  >
    <div
      class="order-none
            
            text-red
            flex items-center w-[1.375rem] text-gray [&>svg]:w-full"
    >
      <svg
        class="tabler-icon tabler-icon-mail"
        data-testid="icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 7a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v10a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2v-10z"
        />
        <path
          d="M3 7l9 6l9 -6"
        />
      </svg>
    </div>
    <input
      class="pl-xs
          
           text-black font-poppins text-base py-xxs bg-transparent border-0 outline-none w-full"
      type="text"
      value=""
    />
  </div>
  <p
    class="text-red text-xs"
  >
    Error message
  </p>
</div>
`;
